### ISSUE TITLE
Add Driver for Apache Calcite

### ISSUE THREAD

**Issue by magno32** [metabase/metabase#6230]
Calcite is a JDBC driver without an underlying data-store.  It provides a means to fairly easily create JDBC access to custom data sources.  

https://calcite.apache.org

There are currently adapters (in the Calcite project itself) that would add support for:

- Cassandra
- CSV (this is their development example)
- Elasticsearch
- Splunk
- More at https://calcite.apache.org/docs/adapter.html

Calcite also has a JDBC adapter, that allows piping JDBC requests through this driver. 

#### Current version is 1.14.0

We are using Calcite to give us a JDBC layer for a proprietary REST api, would will be really nice to have inside Metabase.

**Comment by moumny** [comment]
Could replace the current Druid driver ! 
Druid SQL driver would support make it more inline with other drivers, benefit from query optimizations

**Comment by njhartwell** [comment]
I might look at picking this one up and submitting a PR in the near future. Thoughts / preferences on whether to use the Calcite JDBC driver vs. using http API directly? I've done a lot more work with the latter, but am open to trying the jdbc route if there is a compelling reason.

**Comment by njhartwell** [comment]
Started looking into this today and discovered that druid+calcite doesn't support prepared statements 😞, so this is going to be a lot more work than I had bargained for.

**Comment by moumny** [comment]
I heard they run tableau on top of druid by faking postgres. Maybe its
quite similar?

**Comment by magno32** [comment]
I started looking into this, but could not devote the time to learning Clojure.  I still think this would be a huge add for Metabase.  I would honestly think (though my knowledge is limited) that using the base supported JDBC (ISQLDriverDefaultsMixin) you get a lot of the implementation for free, but I'm not certain.

**Comment by njhartwell** [comment]
I started down the JDBC path and ISQLDriverDefaultsMixin does indeed get you there for free (I don't know much Clojure, but in a couple hours was able to get it to connect to Druid using the calcite/avatica jdbc driver), but then hit the wall of druid not supporting prepared statements. This might not be an issue for all Calcite back ends, but my current need is to get better Druid + Metabase integration so that I can do lookups, parameterized queries, etc. A couple ways forward (all of which look kind of hard from my perspective):
 - Get druid to support prepared statements
 - Make calcite jdbc driver emulate prepared statements
 - Make Metabase support parameterized queries that don't use prepared statements (i.e. interpolate the params and send the query in one piece). I like this option best, since I'm already looking this codebase, but if if the maintainers think it's a bad idea then so do i 😁 .

**Comment by salsakran** [comment]
@njhartwell we're highly unlikely to re-work our current drivers away from prepared statementsjust to support Calcite. 

Not a priority for the core team, and any community PR that does this would need to be gone over with a fine-toothed comb given the huge surface area (13 other DB drivers) for DB specific quirks, bugs and security attack surface area.

**Comment by njhartwell** [comment]
Thanks for the follow-up @salsakran. After thinking about it a bit more I think that making the Calcite driver emulate statements would be less intrusive and is not unheard of in the JDBC world. I'll try going that route next.

**Comment by stijnder** [comment]
Any progress on this @njhartwell?

**Comment by smic-datalabs-von** [comment]
Any updates on this? Thanks!

**Comment by denisesenguel** [comment]
I'm also very interested in that! 

We're using Metabase based on Druid in our team and Druid is currently expanding their SQL query capabilities, especially by introducing join queries. Druid SQL depends on this driver. 

So far Druid and Metabase is not the perfect fit (like it is with Postgres) but with enabling SQL it might just be! Solving this issue will definitely make your Druid users very happy and attract many more

**Comment by smic-datalabs-von** [comment]
@denisesenguel , for the meantime, what dashboard solutions can you suggest that has a nice compatibility with Druid SQL? I'm considering Superset but I haven't tried it yet.

**Comment by denisesenguel** [comment]
@smic-datalabs-von unfortunately not, I also wanted to checkout Superset to see if we can use it for analyses requiring joins until this feature is added to Metabase. If you've found a good solution let me know!

**Comment by dharamgajera** [comment]
> @denisesenguel , for the meantime, what dashboard solutions can you suggest that has a nice compatibility with Druid SQL? I'm considering Superset but I haven't tried it yet.

Check out redash.io !

**Comment by dharamgajera** [comment]
For Druid users, this will be very much required now as Druid has join feature introduced from 0.18.0 version.

**Comment by lkm** [comment]
Druid 0.18 also introduced support for dynamic parameters and prepared statements, so that should no longer be blocking this. See https://github.com/apache/druid/releases/tag/druid-0.18.0 and https://druid.apache.org/docs/0.18.0/querying/sql.html#dynamic-parameters

**Comment by mysticaltech** [comment]
This is really needed folks, even Druid itself recommend using SQL from now on. Superset has also depreciated the native Druid backend in favor of the SQL one.

**Comment by mysticaltech** [comment]
Anyone tried interfacing to druid with a [presto-db](https://prestodb.io/) layer? https://prestodb.io/docs/current/connector/druid.html

**Comment by penelopeatkinson** [comment]
Enabling Druid SQL in Metabase would be a total game-changer for our team too! It would allow far more of your Druid users to use Metabase for more complex analyses. Would love to see this happen!

**Comment by denisesenguel** [comment]
> Anyone tried interfacing to druid with a [presto-db](https://prestodb.io/) layer? https://prestodb.io/docs/current/connector/druid.html

@mysticaltech no! but definitely thinking about it if we can't expect Druid SQL to work in Metabase anytime soon. Would also be interested in other people's experiences :)

**Comment by mysticaltech** [comment]
@denisesenguel Please let us know if you ever do, I may go that route too, but it's a shame really to have to add yet another layer of complexity, for something that I think could be done in a day or two by a dev with the right skills.

**Comment by TheRum** [comment]
@denisesenguel & @mysticaltech We have evaluated redash, metabase and superset for our Druid data visualisation.
And Finally we decided to continue with Superset as it was specifically created for Druid. Also, superset founders have started preset.io company. So, Superset development is now more streamlined with a clear Vision.

**Comment by mysticaltech** [comment]
Thanks for sharing @TheRum! Me and my team are actively considering Superset too. It's well Supported by AirBnB and they use Druid too. The only thing I am going to miss is the beautiful automatic discovery, question interface, and smooth look and feel of Metabase, it would be such a shame!

**Comment by TheRum** [comment]
@mysticaltech Exactly ! Metabase is visually appealing while Superset is feature rich. One has to make a trade-off between these two.

**Comment by mysticaltech** [comment]
Hey all, just wanted to report back that [Presto](https://prestosql.io/), when deployed as a proxy in front of Druid, does make it possible to write SQL queries in Metabase for Druid.

**Comment by denisesenguel** [comment]
@mysticaltech we're also trying out Presto now, thanks for your update! For us it's working super slowly though and somehow Metabase is always timing out after 60secs, and as most Presto queries are taking longer to produce results, this is quite a problem. Any chance you ran into similar issues? 🙏🏾

**Comment by flamber** [comment]
@denisesenguel If the queries run slow on Presto, then you probably need to debug that in a Presto forum. As for Metabase timeout, are you sure it's not your reverse proxy closing the connection to the browser, which means Metabase will stop the query?

**Comment by mysticaltech** [comment]
@denisesenguel For us it works normally. Make sure your system has enough resources, **Presto** is heavy and needs at least 16 GB of RAM. To help you out, here's the presto config I use to make it work between **Druid** and **Metabase**. Good luck 🍀

**config.properties**
```
#single node install config
coordinator=true
node-scheduler.include-coordinator=true
http-server.http.port=8080
discovery-server.enabled=true
discovery.uri=http://presto:8080
```

**jvm.properties**
```
-server
-Xmx16G
-XX:-UseBiasedLocking
-XX:+UseG1GC
-XX:+ExplicitGCInvokesConcurrent
-XX:+HeapDumpOnOutOfMemoryError
-XX:+UseGCOverheadLimit
-XX:+ExitOnOutOfMemoryError
-XX:ReservedCodeCacheSize=512M    
```

**catalog/druid.properties**
```
connector.name=druid
connection-url=jdbc:avatica:remote:url=http://druid-broker:8082/druid/v2/sql/avatica/  
```

**Comment by lkm** [comment]
@mysticaltech This is getting very off topic, but I've found that metabase queries in particular are very slow on large tables through Presto possibly because of the way dates are handled using `date_trunc`. I'm seeing my queries resulting in a full table scan in Presto's druid connector without any filters passed on to Druid, making queries against very large tables impossible. Do you have any ideas on how to mitigate that?

```sql
# This seems to be the problem part
WHERE date_trunc('hour', "druid"."table"."__time")
BETWEEN date_trunc('hour', date_add('hour', -4, now())) AND date_trunc('hour', date_add('hour', -1, now())) LIMIT 2000

# If I run this manually there's no problem (without date_trunc on column)
WHERE "druid"."table"."__time"
BETWEEN date_trunc('hour', date_add('hour', -4, now())) AND date_trunc('hour', date_add('hour', -1, now())) LIMIT 2000
```

**Comment by mysticaltech** [comment]
@lkm I have no idea, the tables I have are small, but maybe you could try setting `compute-pushdown-enabled=true` in catalog/druid.properties. I guess it's worth a shot! Good luck and please let us know! 🍀 More about this setting here https://prestodb.io/docs/current/connector/druid.html

**Comment by lkm** [comment]
Thanks again. I should've been clearer before, the first of the query examples is Metabase generated from the UI and the second is my attempt a fixing it by hand. It seems that Presto doesn't yet support pushing down date_trunc to the database so timestamp based data (which is everything in Druid) would result in a complete table load/scan - ie. Druid would just be a complex data storage system and queries would be terribly slow. This could be the same issue @denisesenguel was experiencing.

I was a little mixed up between the PrestoDB and PrestoSQL documentation, *PrestoSQL* does not yet support aggregation pushdown (or `compute-pushdown-enabled=true`) to Druid although work is under way (see https://github.com/prestosql/presto/issues/4109), however, that would not resolve the issue I'm seeing with the timestamp but it would allow Druid to partipate in query optimization. 

To my mind that voids the Metabase-Presto combination, as I would assume that people would only use Druid for substantial amounts of data and would like to avoid loading entire tables on each query. For our biggest tables (~50-100 million rows) I wasn't able to complete the query because there was a unrelated JSON serialization error in Druid itself after around 6 minutes execution time - and that's on some pretty hefty hardware on 10G SFP network.

I assume there's a good reason why Metabase is wrapping the source column in the `date_trunc` function and not just settles with the between operator but until there's better pushdown in Presto this will kill most of my queries at least. So I'm back to hoping that someone who knows Clojure would convert the Metabase Druid driver to Calcite JDBC.

For completeness this is the explain analyse statement results for the above queries (notice especially 28773 vs 6 rows).
```sql
EXPLAIN ANALYZE SELECT *
FROM "druid"."druid"."test"
WHERE date_trunc('hour', "__time")
BETWEEN date_trunc('hour', date_add('hour', -4, now())) AND date_trunc('hour', date_add('hour', -1, now()))


Fragment 1 [SOURCE]
    CPU: 227.92ms, Scheduled: 762.32ms, Input: 28773 rows (0B); per task: avg.: 28773.00 std.dev.: 0.00, Output: 8 rows (1.61kB)
    Output layout: [__time, name]
    Output partitioning: SINGLE []
    Stage Execution Strategy: UNGROUPED_EXECUTION
    ScanFilter[table = druid:druid.test druid.druid.test, grouped = false, filterPredicate = (("@date_trunc@CUT"('hour', "__time") >= "@$literal$@CUT"(1601690400000000)) AND ("@date_trunc@CUT"('hour', "__time") <= "@$literal$@CUT"(1601701200000000)))]

# VS

Fragment 1 [SOURCE]
    CPU: 29.52ms, Scheduled: 52.76ms, Input: 6 rows (1.20kB); per task: avg.: 6.00 std.dev.: 0.00, Output: 6 rows (1.20kB)
    Output layout: [__time, name]
    Output partitioning: SINGLE []
    Stage Execution Strategy: UNGROUPED_EXECUTION
    TableScan[druid:druid.test druid.druid.test, grouped = false]
```

**Comment by mysticaltech** [comment]
@Ikm you are right, prestosql and prestodb are not the same project, but forks I would guess. However prestodb is also open source and does support pushdown. Maybe try using it instead.

**Comment by mysticaltech** [comment]
Citing [Start Burst Data](https://www.starburstdata.com/prestosql-and-prestodb) 

> While PrestoDB was built to make queries more efficient for companies like Facebook and Uber, PrestoSQL is built for big data exploration for a much broader audience.

@lkm that raises the chance of the pushdown from [PrestoDB](https://prestodb.io/docs/current/connector/druid.html) working for you! 🤞

**Comment by mysticaltech** [comment]
@camsaul Apache Druid is now at the core of many Big Data setups. It would be great if Metabase moved away from the unpractical Druid native query language and started supporting the new and shiny Druid SQL. There are no more blockers to do so, please see below! 🙏🏻 

> Druid 0.18 also introduced support for dynamic parameters and prepared statements, so that should no longer be blocking this. See https://github.com/apache/druid/releases/tag/druid-0.18.0 and https://druid.apache.org/docs/0.18.0/querying/sql.html#dynamic-parameters

This would of course be a lot better than us trying to hack away a heavy **Presto** instance in between! 😢

**Comment by lkm** [comment]
@mysticaltech If only. I had already attempted that, and it's the same result unfortunately. Found this issue which seems to ask for that particular functionality: https://github.com/prestodb/presto/issues/7720

I might do a bit more investigation next week of the practicality of implementing it in PrestoSQL as Java is something I can work with. Presto does offer the possiblity to join across multiple data sources with would be quite interesting for us if we could make it performant.

**Comment by lkm** [comment]
The Presto community is working on that functionality as it's not supported for any database (not just Druid) through Presto so it's a much bigger job. I've instead created an issue for Metabase requesting to change the generated query here https://github.com/metabase/metabase/issues/13361

**Comment by repolevedavaj** [comment]
I started to create a custom Metabase plugin based on Apache Calcite to connect to the Stripe API. In case someone else is working on something similar, I am interested to exchange related to some issues i am facing.

E.g. when running this statement (generated by Metabase):

```
SELECT
  floor(
    TIMESTAMPADD(
      second,
      "stripe"."SUBSCRIPTIONS"."START_DATE",
      timestamp '1970-01-01 00:00:00'
    ) to month
  ) AS "START_DATE",
  COUNT(*) AS "count"
FROM
  "stripe"."SUBSCRIPTIONS"
GROUP BY
  floor(
    TIMESTAMPADD(
      second,
      "stripe"."SUBSCRIPTIONS"."START_DATE",
      timestamp '1970-01-01 00:00:00'
    ) to month
  )
ORDER BY
  floor(
    TIMESTAMPADD(
      second,
      "stripe"."SUBSCRIPTIONS"."START_DATE",
      timestamp '1970-01-01 00:00:00'
    ) to month
  ) ASC
```

I get:

```cannot convert to Object (with type) (org.apache.calcite.avatica.util.AbstractCursor$TimestampFromNumberAccessor@1226926)```

Update: I just realized that Apache Calcite does not support `java.time` classes. With `use-legacy-classes-for-read-and-set` it works!
